doc/manual/metaprogramming.rst

```diff
@@ -538,6 +660,57 @@ is occasionally quite handy.
 
 .. _man-non-standard-string-literals2:
 
+Code Generation
+~~~~~~~~~~~~~~~
+
+When a significant amount of repetitive boilerplate code is required, it
+is common to generate it programmatically to avoid redundancy. In most
+languages, this requires an extra build step, and a separate program to
+generate the repetitive code. In Julia, expression interpolation and
+:func:`eval` allow such code generation to take place in the normal course of
+program execution. For example, the following code defines a series of
+operators on three arguments in terms of their 2-argument forms::
+
+    for op = (:+, :*, :&, :|, :$)
+      eval(quote
+	($op)(a,b,c) = ($op)(($op)(a,b),c)
+      end)
+    end
+
+In this manner, Julia acts as its own preprocessor, and allows code
+generation from inside the language. The above code could be written
+slightly more tersely using the ``:`` prefix quoting form::
+
+    for op = (:+, :*, :&, :|, :$)
+      eval(:(($op)(a,b,c) = ($op)(($op)(a,b),c)))
+    end
+
+This sort of in-language code generation, however, using the
+``eval(quote(...))`` pattern, is common enough that Julia comes with a
+macro to abbreviate this pattern::
+
+    for op = (:+, :*, :&, :|, :$)
+      @eval ($op)(a,b,c) = ($op)(($op)(a,b),c)
+    end
+
+The :obj:`@eval` macro rewrites this call to be precisely equivalent to the
+above longer versions. For longer blocks of generated code, the
+expression argument given to :obj:`@eval` can be a block::
+
+    @eval begin
+      # multiple lines
+    end
+
+Interpolating into an unquoted expression is not supported and will
+cause a compile-time error:
+
+.. doctest::
+
+    julia> $a + b
+    ERROR: unsupported or misplaced expression $
+
```

Comment:
@jakebolewski, care to clarify what you mean by this remark? There are places where you need the second in order to get the dispatch result you want (`Union`s are penalized when it comes to method matching).